include/revng/TupleTree/TupleTreeDiff.h

```diff
@@ -38,15 +44,58 @@ void addToContainer(C &Container, const typename C::value_type &Value) {
   Container.push_back(Value);
 }
 
+template<typename T>
+struct TupleTreeEntries {
+  using Types = std::variant<char>;
+};
+
+template<typename T>
+using TupleTreeEntriesT = typename TupleTreeEntries<T>::Types;
+
+template<>
+struct TupleTreeEntries<model::Binary> {
+  using Types = model::AllTypes;
+};
+
+template<typename T>
+struct Change {
+public:
+  Change() = default;
+  using TupleTreeType = T;
+  explicit Change(TupleTreePath Path,
+                  std::optional<TupleTreeEntriesT<T>> Old,
+                  std::optional<TupleTreeEntriesT<T>> New) :
+    Path(std::move(Path)), Old(std::move(Old)), New(std::move(New)) {}
+
+  static Change removal(TupleTreePath Path, TupleTreeEntriesT<T> Old) {
+    return Change(std::move(Path), std::move(Old), std::nullopt);
+  }
+
+  static Change addition(TupleTreePath Path, TupleTreeEntriesT<T> New) {
+    return Change(std::move(Path), std::nullopt, std::move(New));
+  }
+
+  static Change change(TupleTreePath Path,
+                       TupleTreeEntriesT<T> Old,
+                       TupleTreeEntriesT<T> New) {
+    return Change(std::move(Path), std::move(Old), std::move(New));
+  }
+
+  using EntryType = std::optional<TupleTreeEntriesT<T>>;
+
+  TupleTreePath Path;
+  EntryType Old = std::nullopt;
+  EntryType New = std::nullopt;
+};
```

Comment:
In general, and in future too, please try to put methods/fields in the conventional order we use:

```c++
class MyClass {
private:
  // private usings, if any

public:
  // public usings, if any

private: // Or public
  // Fields of the class

public:
  // Constructors

public:
  // static factory methods

public:
  // Accessor methods

public:
  // Non-accessor `const` methods

public:
  // Non-accessor const methods

private:
  // Helper private methods
};
```
This kind of conventions meets the readers expectations and make him feel comfy.

Also, at least for important data structures, make sure you have a `verify` (is having both fields `nullopt` valid?) and a `dump() debug_function` method.
Future developer debugging this will love you.

I've noted these [here (B.16, B.17)](https://pad.rev.ng/s/SJNUet_ku#).